peps/pep-0748.rst

```diff
@@ -0,0 +1,923 @@
+where a user wants to connect to some IoT device which only has a self-signed
+certificate, it is strongly recommended to add this certificate into a custom
+trust store, rather than using the insecure module to disable certificate
+validation.
+
+Changes to the Standard Library
+===============================
+
+The portions of the standard library that interact with TLS should be revised to
+use these Protocol classes. This will allow them to function with other TLS
+backends. This includes the following modules:
+
+* :mod:`asyncio`
+* :mod:`ftplib`
+* :mod:`http`
+* :mod:`imaplib`
+* :mod:`nntplib`
+* :mod:`poplib`
+* :mod:`smtplib`
+* :mod:`urllib`
+
+Migration of the ssl module
+---------------------------
+
+Naturally, we will need to extend the :mod:`ssl` module itself to conform to
+these Protocol classes. This extension will take the form of new classes,
+potentially in an entirely new module. This will allow applications that take
+advantage of the current :mod:`ssl` module to continue to do so, while enabling
+the new APIs for applications and libraries that want to use them.
+
+In general, migrating from the :mod:`ssl` module to the new Protocol classes is
+not expected to be one-to-one. This is normally acceptable: most tools that use
+the :mod:`ssl` module hide it from the user, and so refactoring to use the new
+module should be invisible.
+
+However, a specific problem comes from libraries or applications that leak
+exceptions from the :mod:`ssl` module, either as part of their defined API or by
+accident (which is easily done). Users of those tools may have written code that
+tolerates and handles exceptions from the :mod:`ssl` module being raised:
+migrating to the protocol classes presented here would potentially cause the
+exceptions defined above to be thrown instead, and existing ``except`` blocks
+will not catch them.
+
+For this reason, part of the migration of the :mod:`ssl` module would require
+that the exceptions in the :mod:`ssl` module alias those defined above. That is,
+they would require the following statements to all succeed:
+
+.. code-block:: python
+
+    assert ssl.SSLError is tls.TLSError
+    assert ssl.SSLWantReadError is tls.WantReadError
+    assert ssl.SSLWantWriteError is tls.WantWriteError
+
+
+The exact mechanics of how this will be done are beyond the scope of this PEP,
+as they are made more complex due to the fact that the current ssl exceptions
+are defined in C code, but more details can be found in `an email sent to the
+Security-SIG by Christian Heimes
+<https://mail.python.org/pipermail/security-sig/2017-January/000213.html>`_.
+
+Future
+======
+
+Major future TLS features may require revisions of these protocol classes. These
+revisions should be made cautiously: many backends may not be able to move
+forward swiftly, and will be invalidated by changes in these protocol classes.
+This is acceptable, but wherever possible features that are specific to
```

Comment:
This sentence is odd. This new module would be expected to follow the same compatibility rules as the rest of the standard library (the existing SSL module exception for updating to more secure defaults over time would apply)

Reply:
Agreed; I think we took this verbatim from PEP 543, so this entire "future" section probably needs to be reworked.